### ISSUE TITLE
consul-template for prepared queries

### ISSUE THREAD

**Issue by bensojona** [hashicorp/consul-template#530]
I don't believe this is currently possible as the feature was just released in Consul 0.6 and introduced a new endpoint (`/v1/query`). But, it would be awesome if you could take advantage of prepared queries to do datacenter failover for service discovery in consul-template.

**Comment by slackpad** [comment]
We've been thinking a bit about what this integration would look like (does CT make the query for you and then tie it to a session for cleanup)? If you avoid trying to define a query via CT, which is probably a good way to go, then you could definitely execute a query and iterate over the results, which would be useful. For now you'd have to use DNS which is pretty limiting.

**Comment by bensojona** [comment]
@slackpad makes sense!

**Comment by slackpad** [comment]
Noting the tricky part is that the [execute endpoint][https://www.consul.io/docs/agent/http/query.html#execute] doesn't support blocking semantics so we'd have to figure that out, either adding blocking which would be very complex given the possible behaviors of failing over, etc. or adding some kind of polling interval on the CT side. Need to think about the better UX there.

**Comment by jippi** [comment]
For my personal case, a per-entry (maybe?) interval for refreshing 

e.g. `{{ with service mysql "near=_agent" "refresh=60s" }}`

**Comment by jippi** [comment]
or maybe poll a unprepared version (e.g. a service), and when that changes, refresh the prepared version

``{{ with service mysql | prepared mysql_locality }} `` would refresh the prepared every time the other service trigger a change?

**Comment by msvechla** [comment]
Is there currently a way for querying services of all datacenters? If I do not give any service, always the local datacenter is used. In my case I need the services from all datacenters. Is there a simple way to do this with CT?

**Comment by slackpad** [comment]
Hi @msvechla there's no first-class way to do this. You range of the list of datacenters via `{{ datacenters }}` and feed that into a service query, though.

**Comment by leprechau** [comment]
@msvechla you probably already figured it out but it takes a bit more finagling with to actually get what you really want.  Something like this should work ... the real trick is the printf to get the format of the datacenter name into something that the ```services``` call will accept.
```
{{range datacenters -}}
	{{- $dcOpt := printf "@%s" . -}}
	{{- range services $dcOpt -}}
		{{.Name}}{{$dcOpt}}{{"\n"}}
	{{- end -}}
{{end}}
```

**Comment by kak-tus** [comment]
@slackpad @sethvargo Hello, can I able to use consul-template now with prepared queries? Main problem is to get port of service in template (ip I can get thru dns with some_service.query.consul).

**Comment by jippi** [comment]
@slackpad is there anything in this ticket i can help with? it's something I would love to use within Nomad template{} - is it blocked on upstream features in Consul or system design in CT ?

**Comment by slackpad** [comment]
Hi @jippi,

> For my personal case, a per-entry (maybe?) interval for refreshing

That approach looks pretty good for this. @sethvargo what do you think?

**Comment by wryfi** [comment]
Is anyone working on this? It would be very nice to have in my environment.

**Comment by lvboudre** [comment]
is this dead? its been 10 years, comon guys, it's lame not to have it.
